From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 21 Apr 2023 21:29:41 +0200
Subject: [PATCH] Only broadcast item pickup to picker viewers

Stops people with modified clients from detecting vanished players based on invalid entities in this packet

diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index 7eaee0d0dcbb420abb5c49ba0a465d90fe513551..641d652527025a4705c27300fc4e99604f78a8c6 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -1139,7 +1139,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
 
     }
 
-    protected void broadcastAndSend(Entity entity, Packet<?> packet) {
+    public void broadcastAndSend(Entity entity, Packet<?> packet) { // CloudPlane - protected -> public
         ChunkMap.TrackedEntity playerchunkmap_entitytracker = (ChunkMap.TrackedEntity) this.entityMap.get(entity.getId());
 
         if (playerchunkmap_entitytracker != null) {
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 418296b14835cf48137b175790c5e62847364a63..17a22b58c7968ce2c1f65ba5d82d0d3321ca9799 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3721,7 +3721,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
     public void take(Entity item, int count) {
         if (!item.isRemoved() && !this.level().isClientSide && (item instanceof ItemEntity || item instanceof AbstractArrow || item instanceof ExperienceOrb)) {
-            ((ServerLevel) this.level()).getChunkSource().broadcast(item, new ClientboundTakeItemEntityPacket(item.getId(), this.getId(), count));
+            ((ServerLevel) this.level()).getChunkSource().chunkMap.broadcastAndSend(this, new ClientboundTakeItemEntityPacket(item.getId(), this.getId(), count)); // CloudPlane - broadcast to pickup viewers
         }
 
     }
